### PR TITLE
added lda ctr recommender

### PR DESCRIPTION
**Pull request by ndomar**

### CODE REVIEW

File: lib/LDA_CTR.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        lambdaI = numpy.eye(self.n_factors) * _lambda
+        for u in range(latent_vectors.shape[0]):
+            """
+            Omar code:
+            confidence = self.build_confidence_matrix(u, 'user')
+            YTY = (fixed_vecs.T * confidence).dot(fixed_vecs)
+            latent_vectors[u, :] = solve((YTY + lambdaI), (ratings[u, :] * confidence).dot(fixed_vecs))
+            """
+            # Anas Code
+            YTY = (fixed_vecs.T * confidence_matrix[u]).dot(fixed_vecs)
```
Comment (mostafa-mahmoud) [inline lib/LDA_CTR.py:113]: What kind of update that we achieve by this loop? I don't see latent_vectors updated.

File: lib/LDA_CTR.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        super(LDACTRRecommender, self).__init__(initializer, evaluator, hyperparameters, options,
+                                             verbose, load_matrices, dump_matrices)
+
+    def train_k_fold(self):
+        all_errors = []
+        for current_k in range(self.k_folds):
+            self.set_data(*self.evaluator.get_fold(current_k, self.fold_test_indices))
+            self.hyperparameters['fold'] = current_k
+            self.train_one_fold(False)
+            all_errors.append(self.get_evaluation_report())
```
Comment (mostafa-mahmoud) [inline lib/LDA_CTR.py:40]: all_errors.append(self.train_one_fold())
should be the same like the last 2 lines

File: lib/LDA_CTR.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        if matrix_found is False:
+            if self._verbose and self._load_matrices:
+                print("Document distribution file was not found, will train LDA.")
+            self._train()
+            if self._dump_matrices:
+                self.initializer.save_matrix(self.document_distribution, 'document_distribution_lda')
+        self.user_vecs = numpy.random.random((self.n_users, self.n_factors))
+        self.training_data_mask = self.evaluator.get_fold_training_data_mask()
+        confidence_matrix = numpy.zeros(self.train_data.shape)
+        confidence_matrix[self.training_data_mask] = self.train_data[self.training_data_mask]
```
Comment (mostafa-mahmoud) [inline lib/LDA_CTR.py]: This part looks good, would be nicer to move 64-70 to a separate small function.
Also would be more precise if we set it as 1 explicitly with the mask (self.training_data_mask & self.train != 0), just in case we had different ratings later on.

File: lib/LDA_CTR.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            if self._verbose and matrix_found:
+                print("Document distribution was set from file, will not train.")
+        if matrix_found is False:
+            if self._verbose and self._load_matrices:
+                print("Document distribution file was not found, will train LDA.")
+            self._train()
+            if self._dump_matrices:
+                self.initializer.save_matrix(self.document_distribution, 'document_distribution_lda')
+        self.user_vecs = numpy.random.random((self.n_users, self.n_factors))
+        self.training_data_mask = self.evaluator.get_fold_training_data_mask()
```
Comment (mostafa-mahmoud) [inline lib/LDA_CTR.py]: Should this function get_fold_training_data_mask be included in the pull request?
